services/device-registry-mongodb/src/test/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedCredentialServiceTest.java

```diff
@@ -208,4 +216,85 @@ public void testUpdateCredentialsFailsForExceededCredentialsPerDeviceLimit(final
+                UUID.randomUUID().toString());
+        final var dto3 = CredentialsDto.forCreation(
+                tenantId,
+                UUID.randomUUID().toString(),
+                List.of(
+                        Credentials.createPasswordCredential("device3a", "secret"),
+                        Credentials.createPSKCredential("device3b", "shared-secret")),
+                UUID.randomUUID().toString());
+        final var dto4 = CredentialsDto.forCreation(
+                UUID.randomUUID().toString(),
+                UUID.randomUUID().toString(),
+                List.of(
+                        Credentials.createPasswordCredential("device1a", "secret"),
+                        Credentials.createPSKCredential("device1b", "shared-secret")),
+                UUID.randomUUID().toString());
+
+        credentialsDao.create(dto1, NoopSpan.INSTANCE.context())
+            .compose(ok -> credentialsDao.create(dto2, NoopSpan.INSTANCE.context()))
+            .compose(ok -> credentialsDao.create(dto3, NoopSpan.INSTANCE.context()))
+            .compose(ok -> credentialsDao.create(dto4, NoopSpan.INSTANCE.context()))
+            .compose(ok -> {
+                final Promise<JsonObject> resultHandler = Promise.promise();
+                final var filter = MongoDbDocumentBuilder.builder()
+                        .withTenantId(tenantId)
+                        .withAuthId("device1a")
+                        .withType(CredentialsConstants.SECRETS_TYPE_HASHED_PASSWORD)
+                        .document();
+                final var commandRight = new JsonObject()
+                        .put("find", "credentials")
+                        .put("batchSize", 1)
+                        .put("singleBatch", true)
+                        .put("filter", filter)
+                        .put("projection", MongoDbBasedCredentialsDao.PROJECTION_CREDS_BY_TYPE_AND_AUTH_ID);
+                final var explain = new JsonObject()
+                        .put("explain", commandRight)
+                        .put("verbosity", "executionStats");
+                mongoClient.runCommand("explain", explain, resultHandler);
+                return resultHandler.future();
+            })
+            .onComplete(ctx.succeeding(result -> {
+                if (LOG.isTraceEnabled()) {
+                    LOG.trace("result:{}{}", System.lineSeparator(), result.encodePrettily());
+                }
+                ctx.verify(() -> {
+                    final var obj = JsonPointer.from("/queryPlanner/winningPlan/inputStage/inputStage")
+                            .queryJson(result);
+                    assertThat(obj).isInstanceOf(JsonObject.class);
+                    final var indexScan = (JsonObject) obj;
+                    assertThat(indexScan.getString("indexName"))
+                        .isEqualTo(MongoDbBasedCredentialsDao.IDX_CREDENTIALS_TYPE_AND_AUTH_ID);
```

Comment:
Good idea to check that the index is actually used! 😄 

Nitpick:

It may also make sense to check the execution stats for "totalKeysExamined" / "totalDocsExamined" being 1. 

At least when using the hint (see https://github.com/eclipse/hono/issues/2792#issuecomment-887266164) one could also observe that using the index isn't sufficient; it is the number of scanned document that counts in the end.

Reply:
I have pushed corresponding changes